lib/eksekuter.rb

```diff
@@ -14,71 +14,76 @@ def initialize logger: nil
+    env, cmd = split_env_and_cmd(args)
+    params = { env: env, cmd: cmd, opts: opts }
+    process_status = spawn_and_get_status(params)
+    assemble_result(params, process_status)
+  end
 
-    params = { env: env, cmd: cmd, opts: opts, block: block }
+  # Like Eksekuter#spawn but the :out and :err options are ignored; instead
+  # the output is attached to the EksekResult object returned.
+  # Returns a EksekResult object which getters :stdout and :stderr containing
+  # the corresponding output of the spawned process.
+  # @return EksekResult
+  def capture *args, **opts
+    env, cmd = split_env_and_cmd(args)
+    out_read, out_write = IO.pipe
+    err_read, err_write = IO.pipe
+    opts2 = opts.merge(out: out_write, err: err_write)
+    params = { env: env, cmd: cmd, opts: opts2 }
+    process_status = spawn_and_get_status(params, streams)
+    assemble_result(params, process_status, streams)
+  end
 
-    process_status = run_process(params, streams)
+  def run *args, **opts
+    env, cmd = split_env_and_cmd(args)
+    capture = opts.delete(:capture) || false
+    set_streams_in_opts streams, opts
+    params = { env: env, cmd: cmd, opts: opts }
+    process_status = spawn_and_get_status(params, streams)
     assemble_result(params, process_status, streams)
   end
 
   private
 
-  def separate_env_and_cmd args
-    env = args[0].is_a?(Hash) ? args.shift : {}
+  def split_env_and_cmd args
+    args2 = args.clone
+    env = args2.first.is_a?(Hash) ? args2.shift : {}
     env = env.each_with_object({}) { |(k, v), o| o[k.to_s] = v }
-    cmd = args.size == 1 ? args.first : args
+    cmd = args2.size == 1 ? args2.first : args2
     [env, cmd]
   end
 
-  def get_out_err_streams capture
-    if capture
-      out_readable, out_writable = IO.pipe
-      err_readable, err_writable = IO.pipe
-      return {
-        out: { readable: out_readable, writable: out_writable },
-        err: { readable: err_readable, writable: err_writable }
-      }
-    end
-
-    {
-      out: { readable: nil, writable: STDOUT },
-      err: { readable: nil, writable: STDERR }
-    }
+  def get_pip
+    out_readable, out_writable = IO.pipe
+    err_readable, err_writable = IO.pipe
+    return [
+      { readable: out_readable, writable: out_writable },
+      { readable: err_readable, writable: err_writable }
+    ]
   end
 
   def set_streams_in_opts streams, opts
     opts[:out] = streams.fetch(:out).fetch(:writable) unless opts[:out]
     opts[:err] = streams.fetch(:err).fetch(:writable) unless opts[:err]
   end
 
-  def run_process params, streams
+  def spawn_and_get_status params
     pid = spawn(params.fetch(:env), *params.fetch(:cmd), params.fetch(:opts))
-    close_streams streams
     _, process_status = Process.wait2(pid)
     process_status
   end
 
-  def close_streams streams
-    out_writable = streams.fetch(:out).fetch(:writable)
-    err_writable = streams.fetch(:err).fetch(:writable)
-    out_writable&.close if out_writable != STDOUT
```

Comment:
On a second thought, I see a problem with this piece of code:

- These streams could have been created by ourselves in `get_out_err_streams`. Closing them is fine.
- If the streams were provided by the user, it's awkward to decide to close them without consulting the user.
- How do we distinguish between streams provided by the user and streams created by us? We have to add moar code?

Reply:
Is this still relevant? The code you are talking about has been removed...

Reply:
Nope!

But it helps explain why I moved the code around in a way that this function is not useful anymore